mm2src/coins/nft.rs

```diff
@@ -601,39 +647,41 @@ where
     Ok(())
 }
 
-async fn get_moralis_nft_list(
-    ctx: &MmArc,
-    chain: &Chain,
-    url: &Url,
-    url_antispam: &Url,
-) -> MmResult<Vec<Nft>, GetNftInfoError> {
+async fn get_moralis_nft_list(ctx: &MmArc, wrapper: &UrlSignWrapper<'_>) -> MmResult<Vec<Nft>, GetNftInfoError> {
     let mut res_list = Vec::new();
+    let chain = wrapper.chain;
     let ticker = chain.to_ticker();
     let conf = coin_conf(ctx, ticker);
     let my_address = get_eth_address(ctx, &conf, ticker, &HDPathAccountToAddressId::default()).await?;
-    let uri_without_cursor = construct_moralis_uri_for_nft(url, &my_address.wallet_address, chain)?;
+    let uri_without_cursor = construct_moralis_uri_for_nft(wrapper.orig_url, &my_address.wallet_address, chain)?;
 
     // The cursor returned in the previous response (used for getting the next page).
     let mut cursor = String::new();
     loop {
-        let uri = format!("{}{}", uri_without_cursor, cursor);
-        let response = send_request_to_uri(uri.as_str()).await?;
+        // Create a new URL instance from uri_without_cursor and modify its query to include the cursor if present
+        let mut uri = uri_without_cursor.clone();
+        if !cursor.is_empty() {
+            uri.set_query(Some(&cursor));
+        }
+        let payload = moralis_payload_str(uri, wrapper.signed_message.clone())?;
+        let response = send_post_request_to_uri(wrapper.orig_url.as_str(), payload).await?;
```

Comment:
We should not change the way we send requests, otherwise this will not work when using endpoints without proxy. I guess you had to do this because you can't send auth payload in the get requests? Did you consider moving this payload to request headers instead of putting it into request body on the proxy side?

Reply:
>  otherwise this will not work when using endpoints without proxy

Hmm even if we decide to send a GET request with the signed message in the header to the proxy layer, we will still need to include a `proxy_auth` parameter (set to true or false) in the NFT request to correctly build the request on the Komodefi side.

I dont think its a good idea to always include the authentication message in the header and allow users to send GET requests directly to some third-party API. I mean, I dont want to provide this user's information in the header or body to anyone if it's not necessary, thats why we should add proxy_auth param and manage this logic.

Therefore, I suggest adding a proxy_auth boolean in the following places, and if it is false, we can send the HTTP GET request directly to the endpoint:
- In the NFT provider enum:
```rust
#[derive(Clone, Deserialize)]
#[serde(tag = "type", content = "info")]
pub enum NftProviderEnum {
    Moralis { url: Url },
}
```
- In NFT wallet RPCs

Anyway, I think both approaches—header vs. body—will require refactoring in both the Komodefi and proxy repositories. Additionally, since the proxy layer supports requests based on feature logic, the proxy will always need to parse and modify incoming requests according to the feature behavior.

Reply:
> Hmm even if we decide to send a GET request with the signed message in the header to the proxy layer, we will still need to include a proxy_auth parameter (set to true or false) in the NFT request to correctly build the request on the Komodefi side.

Don't we have an actiation process for NFTs so we can pass `proxy_auth` only once? Even if we don't have, you can simply add query param to url (something like `xyz.com/some/endpoint?proxy_auth=true`) or even better, add a header `PROXY_AUTH=1` or `PROXY_AUTH=true`.

> I dont think its a good idea to always include the authentication message in the header and allow users to send GET requests directly to some third-party API.

I didn't understand this.


> I mean, I dont want to provide this user's information in the header or body to anyone if it's not necessary

That information doesn't access to the target proxy. Right now, we remove this payload from the body and also drop the certain headers from request before we sending it to proxy.

Reply:
> Don't we have an actiation process for NFTs so we can pass proxy_auth only once?

No, also I think its better to allow user to decide does RPC need to use proxy or not. Actually the same idea is for Urls fields we ask them in every rpc.

> Even if we don't have, you can simply add query param to url (something like xyz.com/some/endpoint?proxy_auth=true) or even better, add a header PROXY_AUTH=1 or PROXY_AUTH=true.

In KomodoDefi RPC if we want to support both functionality: proxy and direct requests to 3rd party API, user needs to highlight do they want to use proxy or not.

```json
curl --url "http://127.0.0.1:7783" --data '{
  "userpass": "'$USERPASS'",
  "method": "enable_eth_with_tokens",
  "mmrpc": "2.0",
  "params": {
    "ticker": "MATIC",
    "mm2": 1,
    "swap_contract_address": "0x9130b257D37A52E52F21054c4DA3450c72f595CE",
    "fallback_swap_contract": "0x9130b257D37A52E52F21054c4DA3450c72f595CE",
    "nodes": [
      {
        "url": "https://polygon-rpc.com"
      }
    ],
    "erc20_tokens_requests": [],
    "nft_req": {
      "provider":{
        "type": "Moralis",
        "info": {
          "url":"http://localhost:6150/nft-test",
          "proxy_auth":true
        }
      }
    }
  }
}'
```


>>> dont think its a good idea to always include the authentication message in the header and allow users to send GET requests directly to some third-party API.
>
> I didn't understand this.
>
>>> I mean, I dont want to provide this user's information in the header or body to anyone if it's not necessary
>
> That information doesn't access to the target proxy. Right now, we remove this payload from the body and also drop the certain headers from request before we sending it to proxy.

I was talking about variant if we dont ask user to provide "proxy_auth" param in Komodefi RPC. It means that we dont know when to send request to proxy and when directly to 3rd party API.
KomoDefi RPC needs to have "proxy_auth" param. If its `true` then we build special request for Komodo Proxy layer, if `false`, then we send general http get request which we were doing all the time previously.

I dont actually see the advantage of moving auth payload to headers. This way we have to check headers for one feature in proxy layer code and check body for other feature.
Proxy layer understands how to manage feature request by `proxy_type: &ProxyType`, so we dont need to over complicate the logic. If we send auth payload message for all features in body, we just need to parse payload from Body using one function
https://github.com/KomodoPlatform/komodo-defi-proxy/blob/d9922eeb44c694c160e2084d8ce0499bd86fde85/src/proxy/mod.rs#L40-L57
```rust
/// Asynchronously generates and organizes payload data from an HTTP request based on the specified proxy type.
/// This function ensures that requests are properly formatted to the correct service,
/// returning a tuple with the modified request and the structured payload.
pub(crate) async fn generate_payload_from_req(
    req: Request<Body>,
    proxy_type: &ProxyType,
) -> GenericResult<(Request<Body>, PayloadData)> {
    match proxy_type {
        ProxyType::Quicknode => {
            let (req, payload) = parse_payload::<QuicknodePayload>(req, false).await?;
            Ok((req, PayloadData::Quicknode(payload)))
        }
        ProxyType::Moralis => {
            let (req, payload) = parse_payload::<MoralisPayload>(req, true).await?;
            Ok((req, PayloadData::Moralis(payload)))
        }
    }
}
```

well if we have new EVM feature with more complex logic for Proxy layer, which cant use `parse_payload`, then we'll implement new parse function for it.

Reply:
if proxy_auth false, then it suppose that user suggests to send request directly to some other API

```
curl --url "http://127.0.0.1:7783" --data '{"userpass":"'$USERPASS'","method":"update_nft","mmrpc":"2.0","params":{"chains":["POLYGON"],"url":"http://other.example.com","url_antispam": "https://nft.antispam.dragonhound.info", "proxy_auth":false}}'
```

Reply:
UPD: I will provide implementation example in separate branch so we can discuss a real code

Reply:
![image](https://github.com/KomodoPlatform/komodo-defi-framework/assets/39852038/3fc1ae10-755b-4d2f-9660-2089050496be)

As far as I see we have nft initialization request, so we can use `proxy_auth` once on the beginning and we no longer need to send it from client side again. 

And if that param is provided as `true` from client, what we should do is just add the auth payload in a header and do not modify anything on the request other than that I mean do not make it POST if it's a GET request, or don't change it's inner values. That makes it more complicated to maintain both KDF and KDP projects. We just want to proxy the requests and KDF doesn't need to change anything on the request other than providing auth values. Previously we only needed to proxy post requests, so I handled the auth payload in the request body, this should be moved to request HEADER now so we can support any type of request (as we can't put data in body for GET requests).

Reply:
> As far as I see we have nft initialization request, so we can use proxy_auth once on the beginning and we no longer need to send it from client side again.

No, we require url and antispam url for each wallet RPC. same for nft init, we just dont need antispam, as we only need thus up to date info.
```rust
/// Represents detailed information about a Non-Fungible Token (NFT).
/// This struct is used to keep info about NFTs owned by user in global Non-Fungible Token.
#[derive(Clone, Debug, Serialize)]
pub struct NftInfo {
    /// The address of the NFT token.
    pub token_address: Address,
    /// The ID of the NFT token.
    #[serde(serialize_with = "serialize_token_id")]
    pub token_id: BigUint,
    /// The blockchain where the NFT exists.
    pub chain: Chain,
    /// The type of smart contract that governs this NFT.
    pub contract_type: ContractType,
    /// The quantity of this type of NFT owned. Particularly relevant for ERC-1155 tokens,
    /// where a single token ID can represent multiple assets.
    pub amount: BigDecimal,
}
```
^ so for each req we need to as proxy_auth.

There is a note: we still need to do the refactoring to init NFT using NFT config from coin conf. Then for `update_nft` instead of ETH coin we will be asking activated NFT

Reply:
> That makes it more complicated to maintain both KDF and KDP projects

If "proxy_auth" false -> use `fn send_request_to_uri`
if "proxy_auth" true -> create payload str and use `fn send_post_request_to_uri`. proxy should handle the moralis feature with route type and payload type

Reply:
> > That makes it more complicated to maintain both KDF and KDP projects
> 
> If "proxy_auth" false -> use `fn send_request_to_uri` if "proxy_auth" true -> create payload str and use `fn send_post_request_to_uri`. proxy should handle the moralis feature with proxy_type and payload type

1) When proxy_auth is true, KDF acts like it's using different service and changes the request method.
2) KDP is taking the POST request and converting it into GET request.

These costs are pure complexities without any benefit. So why doing that?

Reply:
we had long discussion about default urls for nft and covered urls for gas provider a bit, but you can check one of the last comments https://github.com/KomodoPlatform/komodo-defi-framework/pull/2049#discussion_r1502200690

Reply:
> > > That makes it more complicated to maintain both KDF and KDP projects
> > 
> > 
> > If "proxy_auth" false -> use `fn send_request_to_uri` if "proxy_auth" true -> create payload str and use `fn send_post_request_to_uri`. proxy should handle the moralis feature with proxy_type and payload type
> 
>     1. When proxy_auth is true, KDF acts like it's using different service and changes the request method.
> 
>     2. KDP is taking the POST request and converting it into GET request.
> 
> 
> These costs are pure complexities without any benefit. So why doing that?




1. `When proxy_auth is true, KDF acts like it's using different service`

Well yes :D this actually means that request is going to go through the middleware service (from my point of view of course), which is KDP: KDF -> KDP->komodo.moralis.proxy->Mortalis API.

 If `false`, then send request directly to users API (this is the old way we were sending requests before this PR):
 KDF -> 3rd party user's API -> Moralis API. 
If proxy_auth false, user have to insert moralis priv key themself in 3rd party API, as KDF dont do it.

2. `KDP is taking the POST request and converting it into GET request.`

yep, as Moralis feature requires GET reqs.

Ok lets find the compromise, as we have common goal to make a great project, right?

Lets send GET request with payload in header from KDF to KDP if `"proxy_auth":true`. On KDF side I suppose its better to reuse this function adding optional Header:
```rust
pub async fn send_request_to_uri(uri: &str) -> MmResult<Json, GetInfoFromUriError>
```

If `"proxy_auth":false` then send GET request using `send_request_to_uri` directly to url without additional header.

Reply:
Implemented new parse function in KDP https://github.com/KomodoPlatform/komodo-defi-proxy/pull/22/commits/e8f296db9068610699423629fde554faa2d3a632 for features which handle auth signed message in Header

on KDF side added this functionality support for activation process https://github.com/KomodoPlatform/komodo-defi-framework/pull/2100/commits/7882e6ee1225145d1836cdc2842559c7a4d4821b

Will provide `send_request_to_uri` support for NFT Wallet RPCs (get nft list/transfers, update) tomorrow

Reply:
provided proxy auth and 3rd party api support https://github.com/KomodoPlatform/komodo-defi-framework/pull/2100/commits/d03b6d152652f9bc74091d8657109214f16ea4a9

this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2100/commits/ff92a20030e554a75bd85e7184a82fd62296958e reduces some code duplication